scheduler/src/grpc/agent_lifecycle_service.rs

```diff
@@ -20,15 +28,55 @@ impl LifecycleService for AgentLifecycleSvc {
     /// Called by node agents when they request to join the cluster.
     async fn join_cluster(
         &self,
-        _: Request<ConnectionRequest>,
-    ) -> Result<Response<ConnectionResponse>> {
-        todo!();
-        // Ok(Response::new(ConnectionResponse { status_code: 200 }))
+        request: Request<ConnectionRequest>,
+    ) -> Result<Response<ConnectionResponse>, Status> {
+        let agent_id = request.into_inner().id;
+
+        let mut manager = self.node_agent_manager.lock().map_err(|_| {
+            warn!(
+                "Failed to acquire node manager, refusing registration for agent ID {}",
+                agent_id.clone()
+            );
```

Comment:
Please log by using the `event!` macro from `tracing` and specifying any argument as fields:

```suggestion
            event!(
                Level::WARN,
                agent_id,
                "Failed to acquire node manager, refusing registration for agent"
            );
```